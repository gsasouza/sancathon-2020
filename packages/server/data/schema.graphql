input ContactCreateInput {
  businessType: String!
  email: String!
  name: String!
  businessRequirement: String!
  clientMutationId: String
}

type ContactCreatePayload {
  error: String
  clientMutationId: String
}

"""Equipment data"""
type Equipment implements Node {
  """The ID of an object"""
  id: ID!
  name: String
  lastMaintenance: String
  user: User
  removedAt: String
}

"""A connection to a list of items."""
type EquipmentConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [EquipmentEdge]
  count: Int
}

input EquipmentCreateInput {
  name: String!
  clientMutationId: String
}

type EquipmentCreatePayload {
  equipmentEdge: EquipmentEdge
  error: String
  clientMutationId: String
}

"""An edge in a connection."""
type EquipmentEdge {
  """The item at the end of the edge"""
  node: Equipment

  """A cursor for use in pagination"""
  cursor: String!
}

input EquipmentEditInput {
  id: ID!
  lastMaintenance: String!
  clientMutationId: String
}

type EquipmentEditPayload {
  equipmentEdge: EquipmentEdge
  error: String
  clientMutationId: String
}

input EquipmentRemoveInput {
  id: ID!
  clientMutationId: String
}

type EquipmentRemovePayload {
  equipmentEdge: EquipmentEdge
  error: String
  clientMutationId: String
}

type Mutation {
  UserLogin(input: UserLoginInput!): UserLoginPayload
  UserCreate(input: UserCreateInput!): UserCreatePayload
  UserEdit(input: UserEditInput!): UserEditPayload
  UserRemove(input: UserRemoveInput!): UserRemovePayload
  ContactCreate(input: ContactCreateInput!): ContactCreatePayload
  EquipmentCreate(input: EquipmentCreateInput!): EquipmentCreatePayload
  EquipmentEdit(input: EquipmentEditInput!): EquipmentEditPayload
  EquipmentRemove(input: EquipmentRemoveInput!): EquipmentRemovePayload
  ProductCreate(input: ProductCreateInput!): ProductCreatePayload
  ProductEdit(input: ProductEditInput!): ProductEditPayload
  ProductRemove(input: ProductRemoveInput!): ProductRemovePayload
  ProductSign(input: ProductSignInput!): ProductSignPayload
  ProductUnSign(input: ProductUnSignInput!): ProductUnSignPayload
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""Product data"""
type Product implements Node {
  """The ID of an object"""
  id: ID!
  name: String
  description: String
  price: Float
  meHasSigned: Boolean
  removedAt: String
}

"""A connection to a list of items."""
type ProductConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [ProductEdge]
  count: Int
}

input ProductCreateInput {
  name: String!
  description: String!
  price: Float!
  clientMutationId: String
}

type ProductCreatePayload {
  productEdge: ProductEdge
  error: String
  clientMutationId: String
}

"""An edge in a connection."""
type ProductEdge {
  """The item at the end of the edge"""
  node: Product

  """A cursor for use in pagination"""
  cursor: String!
}

input ProductEditInput {
  id: ID!
  description: String
  price: Float
  clientMutationId: String
}

type ProductEditPayload {
  productEdge: ProductEdge
  error: String
  clientMutationId: String
}

input ProductRemoveInput {
  id: ID!
  clientMutationId: String
}

type ProductRemovePayload {
  productEdge: ProductEdge
  error: String
  clientMutationId: String
}

input ProductSignInput {
  id: ID!
  clientMutationId: String
}

type ProductSignPayload {
  productEdge: ProductEdge
  userEdge: UserEdge
  error: String
  clientMutationId: String
}

input ProductUnSignInput {
  id: ID!
  clientMutationId: String
}

type ProductUnSignPayload {
  productEdge: ProductEdge
  userEdge: UserEdge
  error: String
  clientMutationId: String
}

"""The root of all... queries"""
type Query {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node
  me: User
  users(after: String, first: Int, before: String, last: Int, search: String): UserConnection
  equipments(after: String, first: Int, before: String, last: Int, search: String): EquipmentConnection
  products(after: String, first: Int, before: String, last: Int, search: String): ProductConnection
}

"""User data"""
type User implements Node {
  """The ID of an object"""
  id: ID!
  _id: String
  name: String
  username: String
  products: [Product]
  email: String
  removedAt: String
}

"""A connection to a list of items."""
type UserConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [UserEdge]
  count: Int
}

input UserCreateInput {
  username: String!
  email: String!
  name: String!
  password: String!
  clientMutationId: String
}

type UserCreatePayload {
  userEdge: UserEdge
  error: String
  clientMutationId: String
}

"""An edge in a connection."""
type UserEdge {
  """The item at the end of the edge"""
  node: User

  """A cursor for use in pagination"""
  cursor: String!
}

input UserEditInput {
  id: ID!
  email: String
  name: String
  password: String
  clientMutationId: String
}

type UserEditPayload {
  userEdge: UserEdge
  error: String
  clientMutationId: String
}

input UserLoginInput {
  username: String!
  password: String!
  clientMutationId: String
}

type UserLoginPayload {
  token: String
  error: String
  clientMutationId: String
}

input UserRemoveInput {
  id: ID!
  clientMutationId: String
}

type UserRemovePayload {
  userEdge: UserEdge
  error: String
  clientMutationId: String
}
