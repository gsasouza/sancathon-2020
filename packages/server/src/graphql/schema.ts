import MutationType from './type/MutationType';
import QueryType from './type/QueryType';

import { GraphQLSchema } from 'graphql';

const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
  // subscription: SubscriptionType,
});

export default schema;
