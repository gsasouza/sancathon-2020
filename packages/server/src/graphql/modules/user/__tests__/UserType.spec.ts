import {
  clearDbAndRestartCounters,
  connectMongoose,
  createUser,
  disconnectMongoose,
  getContext,
} from '../../../../../test/helpers';
import { toGlobalId } from 'graphql-relay';
import { graphql } from 'graphql';
import schema from '../../../schema';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('Server: UserType', () => {
  it('Should query an user', async () => {
    const user = await createUser();
    // language=GraphQL
    const query = `
      query Q($id: ID!) {
        user: node(id: $id) {
          id
          ... on User {
            id
            username
            name
            email
          }
        }
      }
    `;

    const variables = {
      id: toGlobalId('User', user._id),
    };
    const rootValue = {};
    const context = await getContext({ user: true });
    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.data).toMatchInlineSnapshot(
      {
        user: {
          id: expect.any(String),
        },
      },
      `
      Object {
        "user": Object {
          "email": "user#1@domain.com",
          "id": Any<String>,
          "name": "User#1",
          "username": "user#1",
        },
      }
    `,
    );
  });

  it('Should query a list of users', async () => {
    await createUser();
    await createUser();
    await createUser();
    await createUser();

    // language=GraphQL
    const query = `
      query Q {
        users (first: 10) {
          edges {
            node {
              name
            }
          }
        }
      }
    `;

    const context = await getContext({ user: true });
    const result = await graphql(schema, query, {}, context, {});

    expect(result.data).toMatchSnapshot();
  });

  it('Should search a list of users', async () => {
    await createUser();
    await createUser();
    await createUser({ name: 'search 1' });
    await createUser({ name: 'search 2' });

    // language=GraphQL
    const query = `
      query Q ($search: String!) {
        users (search: $search) {
          edges {
            node {
              name
            }
          }
        }
      }
    `;

    const variables = { search: 'search' };
    const context = await getContext({ user: true });
    const result = await graphql(schema, query, {}, context, variables);

    expect(result.data).toMatchSnapshot();
  });
});
