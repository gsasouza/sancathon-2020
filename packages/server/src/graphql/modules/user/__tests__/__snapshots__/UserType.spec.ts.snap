// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server: UserType Should query a list of users 1`] = `
Object {
  "users": Object {
    "edges": Array [
      Object {
        "node": Object {
          "name": "User#3",
        },
      },
      Object {
        "node": Object {
          "name": "User#2",
        },
      },
      Object {
        "node": Object {
          "name": "User#1",
        },
      },
      Object {
        "node": Object {
          "name": "User#1",
        },
      },
    ],
  },
}
`;

exports[`Server: UserType Should search a list of users 1`] = `
Object {
  "users": Object {
    "edges": Array [
      Object {
        "node": Object {
          "name": "search 2",
        },
      },
      Object {
        "node": Object {
          "name": "search 1",
        },
      },
    ],
  },
}
`;
